File: scamshield-ai/utils/rule_engine.py
"""
Rule-based scam detection engine.

Fast, deterministic, keyword/pattern scoring that runs BEFORE the LLM call.
Serves two purposes:
  1. Instant baseline score with zero network dependency.
  2. Fallback if the Claude API is unavailable or rate-limited — the app
     still returns a usable result instead of failing outright.

This is intentionally simple (string/regex matching) — hackathon-scoped,
not a production NLP pipeline.
"""

import re

# Scam keyword categories. Each category contributes points and, if matched,
# a human-readable reason string. Keywords are matched case-insensitively
# against the raw input text.
KEYWORD_CATEGORIES = {
    "urgency": {
        "keywords": [
            "urgent",
            "immediately",
            "act now",
            "limited offer",
            "expires today",
            "final notice",
            "last chance",
        ],
        "points": 15,
        "reason": "Urgency tactics detected",
    },
    "threats_or_account_action": {
        "keywords": [
            "account suspended",
            "account blocked",
            "bank blocked",
            "account locked",
            "will be deactivated",
            "account suspension",
            "legal action",
        ],
        "points": 20,
        "reason": "Threatening or account-suspension language detected",
    },
    "credential_request": {
        "keywords": [
            "otp",
            "one time password",
            "verify now",
            "verify your account",
            "confirm your password",
            "enter your pin",
            "share your otp",
        ],
        "points": 25,
        "reason": "Suspicious request for OTP, password, or PIN",
    },
    "reward_or_prize": {
        "keywords": [
            "claim prize",
            "claim your prize",
            "you have won",
            "you've won",
            "lottery",
            "congratulations",
            "free gift",
            "cash reward",
        ],
        "points": 30,
        "reason": "Suspicious reward or prize claim detected",
    },
    "payment_request": {
        "keywords": [
            "click here",
            "pay now",
            "processing fee",
            "send payment",
            "wire transfer",
            "gift card",
            "advance fee",
            "pay a",
            "small fee",
        ],
        "points": 25,
        "reason": "Suspicious payment or action request detected",
    },
    "authority_impersonation": {
        "keywords": [
            "income tax department",
            "this is the police",
            "courier department",
            "customs department",
            "from the bank",
            "official notice",
        ],
        "points": 15,
        "reason": "Possible impersonation of an authority or institution",
    },
    "delivery_scam": {
        "keywords": [
            "redelivery fee",
            "delivery failed",
            "package delivery failed",
            "package will be returned",
            "shipment on hold",
            "customs fee",
            "update your delivery address",
            "missed delivery",
        ],
        "points": 25,
        "reason": "Suspicious delivery or shipping fee request detected",
    },
    "job_scam": {
        "keywords": [
            "work from home job",
            "no experience needed",
            "no experience required",
            "registration fee",
            "send your bank details",
            "earn per week",
            "per week",
            "easy money",
            "part time job offer",
        ],
        "points": 25,
        "reason": "Suspicious job offer pattern detected (unrealistic pay, upfront fee, or unsolicited offer)",
    },
}

# Category label heuristics — used to guess a scam "category" from
# whichever keyword groups fired most, before/alongside the LLM's own
# categorization.
CATEGORY_HINTS = {
    "credential_request": "OTP Fraud",
    "reward_or_prize": "Lottery Scam",
    "payment_request": "Delivery Scam",
    "threats_or_account_action": "Phishing",
    "authority_impersonation": "Phishing",
    "urgency": "Phishing",
    "delivery_scam": "Delivery Scam",
    "job_scam": "Job Scam",
}


def run_rule_based_check(text: str) -> dict:
    """
    Scans input text against keyword categories and returns a baseline
    score, suggested category, and matched reasons.

    Returns:
        {
            "score": int (0-100),
            "category": str,
            "reasons": list[str],
            "matched_categories": list[str]
        }
    """
    if not text or not isinstance(text, str):
        return {"score": 0, "category": "Other", "reasons": [], "matched_categories": []}

    normalized = text.lower()

    score = 0
    reasons = []
    matched_categories = []

    for category_name, config in KEYWORD_CATEGORIES.items():
        for keyword in config["keywords"]:
            if keyword in normalized:
                score += config["points"]
                reasons.append(config["reason"])
                matched_categories.append(category_name)
                break  # only count each category once, even if multiple keywords match

    # Extra signal: excessive exclamation marks or all-caps shouting,
    # common in scam messaging.
    if len(re.findall(r"!", text)) >= 3:
        score += 5
        reasons.append("Excessive use of exclamation marks")

    caps_words = re.findall(r"\b[A-Z]{4,}\b", text)
    if len(caps_words) >= 2:
        score += 5
        reasons.append("Excessive use of capital letters")

    # Combo bonus: reward/prize claim + payment request together is the
    # textbook advance-fee fraud pattern ("you won, now pay a fee to claim it").
    if "reward_or_prize" in matched_categories and "payment_request" in matched_categories:
        score += 15
        reasons.append("Classic advance-fee scam pattern: prize claim combined with a payment request")

    # Combo bonus: a "job offer" that also asks for money or bank details
    # upfront is the defining trait of a job scam, not a real employer.
    if "job_scam" in matched_categories and (
        "payment_request" in matched_categories or "credential_request" in matched_categories
    ):
        score += 15
        reasons.append("Job offer combined with a request for money or personal/bank details")

    score = min(score, 100)

    # Guess category from the highest-weighted matched category.
    category = "Other"
    if matched_categories:
        best = max(
            matched_categories,
            key=lambda c: KEYWORD_CATEGORIES[c]["points"],
        )
        category = CATEGORY_HINTS.get(best, "Other")

    return {
        "score": score,
        "category": category,
        "reasons": reasons,
        "matched_categories": matched_categories,
    }

File: src/lib/ai/ruleEngine.ts
export const KEYWORD_CATEGORIES: Record<string, { keywords: string[]; points: number; reason: string }> = {
  urgency: {
    keywords: [
      "urgent", "urgently", "immediately", "act now", "limited offer",
      "expires today", "final notice", "last chance", "deadline",
      "respond within 24 hours", "within 48 hours", "time sensitive",
      "don't delay", "act immediately", "respond asap", "respond immediately",
      "account will be closed", "account will be terminated",
    ],
    points: 15,
    reason: "Urgency tactics used to pressure victims into acting without thinking",
  },

  threats_or_account_action: {
    keywords: [
      "account suspended", "account blocked", "bank blocked", "account locked",
      "will be deactivated", "account suspension", "legal action", "arrested",
      "warrants have been issued", "case filed against you", "cyber crime",
      "your number has been blocked", "sim card blocked", "account will be closed",
      "your device has been", "hacked", "compromised account",
    ],
    points: 20,
    reason: "Threatening or account-suspension language designed to create panic",
  },

  credential_request: {
    keywords: [
      "otp", "one time password", "one-time password", "verify now",
      "verify your account", "confirm your password", "enter your pin",
      "share your otp", "send your otp", "tell me the otp", "what is the otp",
      "give me the code", "verification code", "confirm otp", "enter otp",
      "share the 6 digit", "share the 4 digit", "bank password",
      "net banking password", "internet banking pin",
    ],
    points: 30,
    reason: "Suspicious request for OTP, PIN, or password — legitimate services never ask for this",
  },

  reward_or_prize: {
    keywords: [
      "claim prize", "claim your prize", "you have won", "you've won",
      "lottery", "congratulations you", "free gift", "cash reward",
      "lucky winner", "selected as a winner", "prize money", "jackpot",
      "you are eligible", "claim your reward", "redeem your prize",
      "kbc lottery", "kbc winner", "you have been selected",
    ],
    points: 30,
    reason: "Unsolicited prize or lottery claim — a hallmark of advance-fee fraud",
  },

  payment_request: {
    keywords: [
      "pay now", "processing fee", "send payment", "wire transfer",
      "gift card", "advance fee", "small fee", "nominal fee",
      "registration fee", "activation fee", "delivery charges",
      "customs duty", "tax payment", "release fee", "handling fee",
      "send rs", "send inr", "transfer the amount", "google play card",
      "itunes card", "amazon gift card", "pay via upi",
    ],
    points: 25,
    reason: "Request for an advance fee or payment using untraceable methods",
  },

  authority_impersonation: {
    keywords: [
      "income tax department", "this is the police", "courier department",
      "customs department", "from the bank", "official notice", "rbi",
      "reserve bank of india", "sebi", "government of india", "ministry",
      "narcotics control", "enforcement directorate", "ed officer",
      "cbi officer", "cyber crime department", "i am calling from",
      "call from amazon", "call from microsoft", "call from apple",
    ],
    points: 20,
    reason: "Impersonation of a government authority, bank, or institution",
  },

  delivery_scam: {
    keywords: [
      "redelivery fee", "delivery failed", "package delivery failed",
      "package will be returned", "shipment on hold", "customs fee",
      "update your delivery address", "missed delivery", "parcel seized",
      "fedex", "dhl delivery", "india post parcel", "your package",
      "courier failed", "re-schedule delivery",
    ],
    points: 20,
    reason: "Fake delivery notification requesting payment or personal details",
  },

  job_scam: {
    keywords: [
      "work from home job", "no experience needed", "no experience required",
      "send your bank details", "earn per week", "easy money",
      "part time job offer", "earn rs daily", "earn per day",
      "simple typing job", "data entry job", "earn from mobile",
      "join our team", "whatsapp job", "telegram job", "pay after joining",
      "you have been selected for", "interview call letter",
    ],
    points: 25,
    reason: "Fake job offer with unrealistic pay or upfront fee requirement",
  },

  crypto_scam: {
    keywords: [
      "bitcoin", "crypto investment", "cryptocurrency", "guaranteed returns",
      "double your investment", "100% returns", "200% returns",
      "trading bot", "forex trading", "binance", "coinbase", "crypto wallet",
      "send bitcoin", "send eth", "ethereum investment", "defi returns",
      "nft giveaway", "crypto airdrop",
    ],
    points: 25,
    reason: "Cryptocurrency or investment scam pattern with promises of guaranteed returns",
  },

  investment_scam: {
    keywords: [
      "guaranteed profit", "guaranteed return", "risk free investment",
      "risk-free investment", "safe investment", "high returns",
      "monthly returns", "weekly returns", "profit guaranteed",
      "sebi registered", "rbi approved", "government scheme",
      "mutual fund scam", "stock tip", "insider tip", "penny stock",
      "pig butchering", "wealth management team",
    ],
    points: 25,
    reason: "Investment scam promising guaranteed or unrealistic returns",
  },

  upi_fraud: {
    keywords: [
      "upi", "upi id", "google pay", "gpay", "phonepe", "paytm",
      "scan qr", "scan qr code", "qr code payment", "collect request",
      "approve the payment request", "upi payment request",
      "enter upi pin", "enter mpin", "upi pin", "upi scam",
      "payment request sent", "approve to receive money",
    ],
    points: 25,
    reason: "UPI fraud attempt — scammers send collect requests disguised as money transfer",
  },

  bank_impersonation: {
    keywords: [
      "your bank account", "sbi", "hdfc", "icici", "axis bank",
      "kotak bank", "yes bank", "bank of india", "union bank",
      "kyc update", "kyc verification", "kyc pending", "kyc expired",
      "link your aadhaar", "aadhaar link", "pan card update",
      "net banking", "internet banking", "mobile banking alert",
    ],
    points: 20,
    reason: "Bank impersonation — fake KYC or account alerts from supposed bank officials",
  },

  romance_scam: {
    keywords: [
      "i love you", "we have a connection", "i am a soldier",
      "i am stationed", "send me money", "help me transfer",
      "oil rig", "inheritance money", "i need your help",
      "dating site", "met online", "let's move to whatsapp",
    ],
    points: 20,
    reason: "Romance or relationship scam indicators detected",
  },
};

const CATEGORY_HINTS: Record<string, string> = {
  credential_request: "OTP Fraud",
  reward_or_prize: "Lottery Scam",
  payment_request: "Delivery Scam",
  threats_or_account_action: "Phishing",
  authority_impersonation: "Phishing",
  urgency: "Phishing",
  delivery_scam: "Delivery Scam",
  job_scam: "Job Scam",
  crypto_scam: "Investment Scam",
  investment_scam: "Investment Scam",
  upi_fraud: "UPI Fraud",
  bank_impersonation: "Phishing",
  romance_scam: "Romance Scam",
};

export function runRuleBasedCheck(text: string) {
  if (!text || typeof text !== "string") {
    return { score: 0, category: "Other", reasons: [], matched_categories: [] };
  }

  const normalized = text.toLowerCase();
  let score = 0;
  const reasons: string[] = [];
  const matched_categories: string[] = [];

  for (const [categoryName, config] of Object.entries(KEYWORD_CATEGORIES)) {
    for (const keyword of config.keywords) {
      if (normalized.includes(keyword)) {
        score += config.points;
        if (!reasons.includes(config.reason)) {
          reasons.push(config.reason);
        }
        if (!matched_categories.includes(categoryName)) {
          matched_categories.push(categoryName);
        }
        break;
      }
    }
  }

  // Stylistic red flags
  const exclamationMatches = text.match(/!/g);
  if (exclamationMatches && exclamationMatches.length >= 3) {
    score += 5;
    reasons.push("Excessive use of exclamation marks creates artificial urgency");
  }

  const capsWords = text.match(/\b[A-Z]{4,}\b/g);
  if (capsWords && capsWords.length >= 2) {
    score += 5;
    reasons.push("Excessive use of capital letters — a pressure tactic");
  }

  // Combo bonuses for compound scam patterns
  if (matched_categories.includes("reward_or_prize") && matched_categories.includes("payment_request")) {
    score += 20;
    reasons.push("Classic advance-fee fraud: prize claim combined with a payment request");
  }

  if (matched_categories.includes("job_scam") &&
    (matched_categories.includes("payment_request") || matched_categories.includes("credential_request"))) {
    score += 15;
    reasons.push("Fraudulent job offer combined with a request for money or personal details");
  }

  if (matched_categories.includes("upi_fraud") && matched_categories.includes("urgency")) {
    score += 15;
    reasons.push("Urgent UPI payment request — scammers rush victims to approve collect requests");
  }

  if (matched_categories.includes("crypto_scam") && matched_categories.includes("investment_scam")) {
    score += 15;
    reasons.push("Cryptocurrency investment scam with guaranteed returns — a common pig-butchering tactic");
  }

  if (matched_categories.includes("authority_impersonation") && matched_categories.includes("threats_or_account_action")) {
    score += 20;
    reasons.push("Official-sounding threats from impersonated authorities — a classic intimidation scam");
  }

  score = Math.min(score, 100);

  let category = "Other";
  if (matched_categories.length > 0) {
    const best = matched_categories.reduce((a, b) =>
      KEYWORD_CATEGORIES[a].points > KEYWORD_CATEGORIES[b].points ? a : b
    );
    category = CATEGORY_HINTS[best] || "Other";
  }

  return {
    score,
    category,
    reasons,
    matched_categories,
  };
}
